server/task.go

```diff
@@ -246,6 +249,39 @@ func (s *Server) registerTaskRoutes(g *echo.Group) {
 		}
 		return nil
 	})
+
+	g.POST("/pipeline/:pipelineID/task/:taskID/cancel", func(c echo.Context) error {
+		ctx := c.Request().Context()
+		taskID, err := strconv.Atoi(c.Param("taskID"))
+		if err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Task ID is not a number: %s", c.Param("taskID"))).SetInternal(err)
+		}
+		task, err := s.store.GetTaskByID(ctx, taskID)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to cancel the task").SetInternal(err)
+		}
+		if task == nil {
+			return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("Task not found with ID %d", taskID))
+		}
+
+		if !taskCancellationImplemented[task.Type] {
+			return echo.NewHTTPError(http.StatusNotImplemented, fmt.Sprintf("Canceling task type %s is not supported", task.Type))
+		}
+		// TODO(p0ny): support cancel pending task.
+		if !isTaskStatusTransitionAllowed(task.Status, api.TaskCanceled) {
+			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid task status transition from %v to %v.", task.Status, api.TaskCanceled))
+		}
+
+		s.TaskScheduler.runningExecutorsMutex.Lock()
+		cancel, ok := s.TaskScheduler.runningExecutorsCancel[taskID]
+		s.TaskScheduler.runningExecutorsMutex.Unlock()
+		if !ok {
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to cancel task")
+		}
+		cancel()
```

Comment:
We should wrap them in Lock()/Unlock().

If two clients concurrently call the cancel API, they could both get ok = true, and call cancel() concurrently.

Reply:
It is safe to do that.
> A CancelFunc may be called by multiple goroutines simultaneously. 

https://pkg.go.dev/context#CancelFunc